src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4328,3 +4328,137 @@ void C2_MacroAssembler::vector_maskall_operation32(KRegister dst, Register src,
   kunpckdql(dst, tmp, tmp);
 }
 #endif
+
+void C2_MacroAssembler::udivI(Register rax, Register divisor, Register rdx) {
+  Label done;
+  Label neg_divisor_fastpath;
+  cmpl(divisor, 0);
+  jccb(Assembler::less, neg_divisor_fastpath);
+  xorl(rdx, rdx);
+  divl(divisor);
+  jmp(done);
+  bind(neg_divisor_fastpath);
+  // Fastpath for divisor < 0:
+  // quotient = (dividend & ~(dividend - divisor)) >>> (Integer.SIZE - 1)
+  // See Hacker's Delight (2nd ed), section 9.3 which is implemented in java.lang.Long.divideUnsigned()
+  movl(rdx, rax);
+  subl(rdx, divisor);
+  andnl(rax, rdx, rax);
```

Comment:
andn needs BMI1 so a check for VM_Version::supports_bmi1() is needed for the intrinsic.